### PR TITLE
Return creating node with (+) button & fix a regression

### PR DESCRIPTION
**Pull request by vitvakatu**
### Pull Request Description

This PR returns previously reverted #3278 PR and also fixes a regression introduced by that PR.

The PR is quite big, so per-commit review is preferred.

The regression consists of two separate issues: 

##### Issue 1
Node searcher thinks that adding a new node by internal API is the same as adding nodes by the user. After the project is loaded, the searcher thinks that one of the project nodes was just created due to an invalid state of the `adding_new_node` FRP output. After the user cancels node editing the searcher removes the "latest added" node.

To fix this bug I added an additional filter for the `should_edit` variable in ProjectView: now `adding_new_node` is only triggered when the node is added by the user, not by an internal API.

##### Issue 2
The second issue is much harder to debug. Let's look at how the searcher opens on the `develop` branch: 
1. `node_being_edited` endpoint is triggered. ProjectView handles this output and the Searcher Presenter is initialized.
2. `node_editing_started` endpoint is triggered. GraphEditor calls `node.model.input.frp.set_edit_mode(true)`, the implementation of `node::input::Area` calls `model.label.set_content` (see `Cursor Setup` section of the `node/input/area.rs`).
3. In turn, this leads to the `graph.node_expression_set` output trigger, which is handled by Searcher Presenter by calling `Searcher::input_changed`.
4. After a bunch of function calls, the Searcher gets an updated action list. Captured network communication between IDE and Engine shows that the searcher asks the action list twice: once after the searcher presenter is initialized and a second time when this `set_content` update happens. Only the second answer returns a non-empty action list for some nodes (like the `.sort` node from the default project). I think this additional communication requires special attention, but I don't know whether this is normal behavior.

Now, let's see what happened in the `(+) button branch`:
1. `node_being_edited` endpoint is triggered. However, this does not initialize Searcher Presenter - it is now initialized later.
2. `node_editing_started` endpoint is triggered. The flow is almost identical, with one important difference: we don't have Searcher Presenter ready yet. Therefore we don't get the second update from the Engine after `node.input.set_content` -> `graph.node_expression_set` call. For some nodes, this leads to an empty list of actions, and the searcher is virtually invisible.
3. ProjectView initializes the Searcher Presenter, but it is too late.

To sum it up: the Engine returns an empty action list for some scene nodes (I reproduced it for `.sort` node of the default project). Accidentally or not, the IDE asks the Engine twice - and it returns a correct action list for the second request. This second request is made by Searcher Presenter after we update the edited node expression (by removing non-connected port names I believe). Therefore the Searcher Presenter must be initialized at a time of `node_editing_started` handling in the Graph Editor.

To fix this bug I changed the used FRP endpoint from `node_editing_started` to `node_being_edited` – and the Searcher Presenter is initialized earlier now.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH `./run dist` and `./run watch`.

### CODE REVIEW

**Comment by wdanilo** [comment]
Ok, so there are few questions that need to be answered here:
1. Why the current implementation asks engine about this information twice? Was that on purpose? Is that documented somewhere?
2. Why engine sometimes returns empty list of entires? If the entries are not filled out yet, it should return information about lack of info, rather than empty list.
3. Regarding if the action was done by the user or was done by the system – I want to chat about it with you and Adam. This is more complex topic that we've been already covering in the past.

@vitvakatu Please check out the points 1 and 2 - talk with people that you need to talk and let's talk about the whole PR after you know the answers.

**Review by wdanilo — APPROVED** [review]

File: app/gui/view/graph-editor/src/lib.rs [hunk 1]
```diff
@@ -398,6 +401,17 @@ impl<K, V, S> SharedHashMap<K, V, S> {
...
 // =================
 
+/// The information about data source hinted by node creation process. For example, when creating
+/// node by dropping edge, the source port should be a source for newly created node.
+///
+/// This is information meant to be sent to searcher, which can, for example, auto- connect the
+/// source to "this" port of new node.
+#[derive(Clone, CloneRef, Copy, Debug, Default, Eq, PartialEq)]
+pub struct NodeSource {
+    #[allow(missing_docs)] // FIXME[everyone] Public-facing API should be documented.
```
Comment (MichaelMauderer) [inline app/gui/view/graph-editor/src/lib.rs]: This FIXME should be addressed or removed.

File: app/gui/view/graph-editor/src/lib.rs [hunk 2]
```diff
@@ -1270,25 +1295,155 @@ impl Deref for GraphEditorModelWithNetwork {
     }
 }
 
+
+impl GraphEditorModelWithNetwork {
+    #[allow(missing_docs)] // FIXME[everyone] Public-facing API should be documented.
```
Comment (MichaelMauderer) [inline app/gui/view/graph-editor/src/lib.rs]: This FIXME should be addressed or removed.

File: app/gui/view/graph-editor/src/lib.rs [hunk 3]
```diff
@@ -1270,25 +1295,155 @@ impl Deref for GraphEditorModelWithNetwork {
...
+                            return true;
+                        }
+                    }
+                }
+            }
+        }
+        false
+    }
+
+    #[allow(missing_docs)] // FIXME[everyone] Public-facing API should be documented.
```
Comment (MichaelMauderer) [inline app/gui/view/graph-editor/src/lib.rs]: This FIXME should be addressed or removed.

File: app/gui/view/graph-editor/src/lib.rs [hunk 4]
```diff
@@ -1610,6 +1697,75 @@ impl GraphEditorModel {
 }
 
 
+// === Add node ===
+impl GraphEditorModel {
+    #[allow(missing_docs)] // FIXME[everyone] Public-facing API should be documented.
```
Comment (MichaelMauderer) [inline app/gui/view/graph-editor/src/lib.rs]: This FIXME should be addressed or removed.

File: app/gui/view/graph-editor/src/lib.rs [hunk 5]
```diff
@@ -1610,6 +1697,75 @@ impl GraphEditorModel {
...
+    pub fn add_node(&self) -> NodeId {
+        self.frp.add_node.emit(());
+        let (node_id, _, _) = self.frp.node_added.value();
+        node_id
+    }
+
+    #[allow(missing_docs)] // FIXME[everyone] Public-facing API should be documented.
```
Comment (MichaelMauderer) [inline app/gui/view/graph-editor/src/lib.rs]: This FIXME should be addressed or removed.

File: app/gui/view/graph-editor/src/lib.rs [hunk 6]
```diff
@@ -1610,6 +1697,75 @@ impl GraphEditorModel {
...
+    pub fn add_node_below(&self, above: NodeId) -> NodeId {
+        let pos = self.find_free_place_under(above);
+        self.add_node_at(pos)
+    }
+
+    #[allow(missing_docs)] // FIXME[everyone] Public-facing API should be documented.
```
Comment (MichaelMauderer) [inline app/gui/view/graph-editor/src/lib.rs]: This FIXME should be addressed or removed.

File: app/gui/view/graph-editor/src/lib.rs [hunk 7]
```diff
@@ -1610,6 +1697,75 @@ impl GraphEditorModel {
...
+    pub fn add_node_at(&self, pos: Vector2) -> NodeId {
+        let node_id = self.add_node();
+        self.frp.set_node_position((node_id, pos));
+        node_id
+    }
+
+    #[allow(missing_docs)] // FIXME[everyone] Public-facing API should be documented.
```
Comment (MichaelMauderer) [inline app/gui/view/graph-editor/src/lib.rs]: This FIXME should be addressed or removed.

File: app/gui/view/graph-editor/src/lib.rs [hunk 8]
```diff
@@ -1610,6 +1697,75 @@ impl GraphEditorModel {
...
+    pub fn find_free_place_under(&self, node_above: NodeId) -> Vector2 {
+        let above_pos = self.node_position(node_above);
+        let y_gap = self.frp.default_y_gap_between_nodes.value();
+        let y_offset = y_gap + node::HEIGHT;
+        let starting_point = above_pos - Vector2(0.0, y_offset);
+        let direction = Vector2(-1.0, 0.0);
+        self.find_free_place_for_node(starting_point, direction).unwrap()
+    }
+
+    #[allow(missing_docs)] // FIXME[everyone] Public-facing API should be documented.
```
Comment (MichaelMauderer) [inline app/gui/view/graph-editor/src/lib.rs]: This FIXME should be addressed or removed.

File: app/gui/view/graph-editor/src/lib.rs [hunk 9]
```diff
@@ -1610,6 +1697,75 @@ impl GraphEditorModel {
...
+            let left = position.x - x_gap - min_spacing;
+            let right = position.x + node.view.model.width() + x_gap;
+            let top = position.y + node::HEIGHT + y_gap;
+            let bottom = position.y - node::HEIGHT - y_gap;
+            OccupiedArea { x1: left, x2: right, y1: top, y2: bottom }
+        });
+        find_free_place(starting_from, direction, node_areas)
+    }
+
+    #[allow(missing_docs)] // FIXME[everyone] Public-facing API should be documented.
```
Comment (MichaelMauderer) [inline app/gui/view/graph-editor/src/lib.rs]: This FIXME should be addressed or removed.

**Review by MichaelMauderer — APPROVED** [review]

File: app/gui/view/graph-editor/src/free_place_finder.rs [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+            direction:       Vector2(1.0, 0.0),
+            occupied:        vec![OccupiedArea { x1: 0.0, x2: 20.0, y1: 15.0, y2: 5.0 }],
+            expected_result: Vector2(20.0, 10.0),
+        };
+        let non_orthogonal = Case {
+            direction: Vector2(18.0, 15.0),
+            expected_result: Vector2(18.0, 15.0),
+            ..orthogonal.clone()
+        };
+        orthogonal.run_each_flip();
```
Comment (akavel) [inline app/gui/view/graph-editor/src/free_place_finder.rs:235]: Are we sure this test is being run by CI and passing successfully? When I selectively run it locally, it appears to fail, as demonstrated in the transcript below. (Note: the patch adding `features = ["serde"]` is a workaround for #3319.)

```console
$ git checkout origin/wip/vitvakatu/fix-regression-181528359
HEAD is now at 278c6c704 Fix regression #181528359

$ patch -p1 <<'EOF'
diff --git a/app/gui/view/graph-editor/Cargo.toml b/app/gui/view/graph-editor/Cargo.toml
index 760d8b24b..07fb743d9 100644
--- a/app/gui/view/graph-editor/Cargo.toml
+++ b/app/gui/view/graph-editor/Cargo.toml
@@ -15,7 +15,7 @@ bimap = { version = "0.4.0" }
 enso-config = { version = "0.1.0", path = "../../config" }
 enso-frp = { version = "0.1.0", path = "../../../../lib/rust/frp" }
 enso-logger = { path = "../../../../lib/rust/logger"}
-enso-prelude = { path = "../../../../lib/rust/prelude"}
+enso-prelude = { path = "../../../../lib/rust/prelude", features = ["serde"] }
 engine-protocol = { version = "0.1.0", path = "../../controller/engine-protocol" }
 enso-shapely = { path = "../../../../lib/rust/shapely"}
 enso-text = { version = "0.1.0", path = "../../../../lib/rust/text" }
EOF
patching file app/gui/view/graph-editor/Cargo.toml

$ RUST_BACKTRACE=1 cargo test free_place_finder --manifest-path=app/gui/view/graph-editor/Cargo.toml
(...)
running 4 tests
test free_place_finder::tests::zero_direction_is_not_allowed ... ok
test free_place_finder::tests::already_in_unoccupied ... ok
test free_place_finder::tests::overlapping_areas ... ok
test free_place_finder::tests::single_area ... FAILED

failures:

---- free_place_finder::tests::single_area stdout ----
thread 'free_place_finder::tests::single_area' panicked at 'assertion failed: `(left == right)`
  left: `Matrix { data: [[-15.0, 15.0]] }`,
 right: `Matrix { data: [[-20.0, 10.0]] }`: Case Case { starting_point: Matrix { data: [[-10.0, 10.0]] }, direction: Matrix { data: [[-10.0, 10.0]] }, occupied: [OccupiedArea { x1: -0.0, y1: 15.0, x2: -20.0, y2: 5.0 }], expected_result: Matrix { data: [[-20.0, 10.0]] } } gave wrong result.', app/gui/view/graph-editor/src/free_place_finder.rs:163:13
stack backtrace:
   0: rust_begin_unwind
             at /rustc/5e57faa78aa7661c6000204591558f6665f11abc/library/std/src/panicking.rs:577:5
   1: core::panicking::panic_fmt
             at /rustc/5e57faa78aa7661c6000204591558f6665f11abc/library/core/src/panicking.rs:110:14
   2: core::panicking::assert_failed_inner
             at /rustc/5e57faa78aa7661c6000204591558f6665f11abc/library/core/src/panicking.rs:186:23
   3: core::panicking::assert_failed
             at /rustc/5e57faa78aa7661c6000204591558f6665f11abc/library/core/src/panicking.rs:149:5
   4: ide_view_graph_editor::free_place_finder::tests::Case::run
             at ./src/free_place_finder.rs:163:13
   5: ide_view_graph_editor::free_place_finder::tests::Case::run_each_flip
             at ./src/free_place_finder.rs:189:13
   6: ide_view_graph_editor::free_place_finder::tests::single_area
             at ./src/free_place_finder.rs:235:9
   7: ide_view_graph_editor::free_place_finder::tests::single_area::{{closure}}
             at ./src/free_place_finder.rs:223:5
   8: core::ops::function::FnOnce::call_once
             at /rustc/5e57faa78aa7661c6000204591558f6665f11abc/library/core/src/ops/function.rs:227:5
   9: core::ops::function::FnOnce::call_once
             at /rustc/5e57faa78aa7661c6000204591558f6665f11abc/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    free_place_finder::tests::single_area

test result: FAILED. 3 passed; 1 failed; 0 ignored; 0 measured; 1 filtered out; finished in 0.06s

error: test failed, to rerun pass '--lib'
```

Comment (vitvakatu) [reply app/gui/view/graph-editor/src/free_place_finder.rs:235]: Fixed! This was a mistake in test implementation + a mistake in expected value. The algorithm works correctly though (as far as I can see)

**Pull request merged**